Allows for stringification of the host class

movermeyer opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
First and foremost, this removes the possibility of `__repr__` returning None (Fixes #24545).

Secondly it changes `__repr__` to `__str__` since the returned string has always been a casual representation of the host, which fits the [definition of `__str__`](https://docs.python.org/2/reference/datamodel.html#object.__str__) and not the  [definition of `__repr__`](https://docs.python.org/2/reference/datamodel.html#object.__repr__)

No Python code should be using the output of `__repr__` or `__str__` in their logic, so this change will not break functionality. In the off-chance that they were, they can easily change to use either `get_name()` or the `name` attribute directly, which will return the same string functionality they've been expecting.

If there is code that relying on the TypeError raised by this issue, then may the [Flying Spaghetti Monster](https://en.wikipedia.org/wiki/Flying_Spaghetti_Monster) have mercy on them.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
Ansible

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### ADDITIONAL INFORMATION
See #24545 for more details.

movermeyer committed:
`__repr__()` can no longer return None 

Fixes #24545

movermeyer committed:
Change `__repr__` to `__str__`

`__repr__` is supposed to be an [unambiguous string representation](https://docs.python.org/2/reference/datamodel.html#object.__repr__). Only including the hostname is better as a `__str__`, which is meant for informal representations.

mattclay commented:
CI failure in integration tests on multiple platforms. Here's one of the failures:

```
2017-05-12 20:05:59 TASK [Gathering Facts] *********************************************************
2017-05-12 20:05:59 ERROR! Unexpected Exception: 'ascii' codec can't encode characters in position 5-9: ordinal not in range(128)
2017-05-12 20:05:59 the full traceback was:
2017-05-12 20:05:59 
2017-05-12 20:05:59 Traceback (most recent call last):
2017-05-12 20:05:59   File "/root/ansible/bin/ansible-playbook", line 109, in <module>
2017-05-12 20:05:59     exit_code = cli.run()
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/cli/playbook.py", line 158, in run
2017-05-12 20:05:59     results = pbex.run()
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/executor/playbook_executor.py", line 153, in run
2017-05-12 20:05:59     result = self._tqm.run(play=play)
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/executor/task_queue_manager.py", line 285, in run
2017-05-12 20:05:59     play_return = strategy.run(iterator, play_context)
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/plugins/strategy/linear.py", line 284, in run
2017-05-12 20:05:59     results += self._process_pending_results(iterator, max_passes=max(1, int(len(self._tqm._workers) * 0.1)))
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 387, in _process_pending_results
2017-05-12 20:05:59     iterator.mark_host_failed(original_host)
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/executor/play_iterator.py", line 492, in mark_host_failed
2017-05-12 20:05:59     display.debug("marking host %s failed, current state: %s" % (host, s))
2017-05-12 20:05:59   File "/root/ansible/lib/ansible/inventory/host.py", line 97, in __str__
2017-05-12 20:05:59     return "Host(name='{0}')".format(self.get_name())
2017-05-12 20:05:59 UnicodeEncodeError: 'ascii' codec can't encode characters in position 5-9: ordinal not in range(128)
```

movermeyer committed:
Update host.py

Switched to a simpler `__str__()` after seeing where it was being used elsewhere. ([Example](https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/play_iterator.py#L492) 

Still not sure how to handle the Unicode problems.

movermeyer commented:
@mattclay: Sorry, but I'm going to need some help on this this one.

I've been fortunate enough to not have to worry about Unicode for much of my development, so I'm not sure of what the patterns are for handling this.

This [StackOverflow](http://stackoverflow.com/a/1307210), seems to be the answer, and it seems to be  [implemented in Ansible here](https://github.com/ansible/ansible/blob/2fff690caab6a1c6a81973f704be3fbd0bde2c2f/lib/ansible/module_utils/six/_six.py#L830), but no one seems to be using it directly. Further, scanning through other definitions of `__str__` in other classes, I don't see any explicit handling for Unicode.

Some help with the proper Unicode patterns would be much appreciated.

mattclay commented:
@movermeyer There is a major overhaul of inventory that is nearing completion which may address the issue you're working on: https://github.com/ansible/ansible/pull/23001/files#diff-d7f638847682d8fca1e1fc96a8084719R49

bcoca commented:
This does not fix an actual bug in our code but use by 3rd party that is not giving needed information to the object, so I'm going to reject this PR.
